_docs/migration_guides/upgrading_to_terragrunt_0.19.x.md

```diff
@@ -0,0 +1,268 @@
+to Terragrunt 0.19.x and newer:
+
+1. [Move from terraform.tfvars to terragrunt.hcl](#move-from-terraformtfvars-to-terragrunthcl)
+1. [Move input variables into inputs](#move-input-variables-into-inputs)
+1. [Use first-class expressions](#use-first-class-expressions)
+1. [Check attributes vs blocks usage](#check-attributes-vs-blocks-usage)
+1. [Rename a few built-in functions ](#rename-a-few-built-in-functions)
+
+   
+### Move from terraform.tfvars to terragrunt.hcl
+
+Since Terraform 0.12 has more strict rules about what can go into `terraform.tfvars` files, you now need to move your
+Terragrunt configuration from `terraform.tfvars` to a `terragrunt.hcl` file, removing the `terragrunt = { ... }`
+wrapping along the way. 
+
+For example, if you had the following in `terraform.tfvars`:
+
+```hcl
+# terraform.tfvars
+
+terragrunt = {
+  terraform {
+    source = "git::git@github.com:foo/modules.git//frontend-app?ref=v0.0.3"
+
+    extra_arguments "custom_vars" {
+      commands  = ["apply", "plan"]
+      arguments = ["-var", "foo=42"]
+    }
+  }
+  
+  remote_state {
+    backend = "s3"
+    config = {
+      bucket         = "my-terraform-state"
+      key            = "${path_relative_to_include()}/terraform.tfstate"
+      region         = "us-east-1"
+      encrypt        = true
+      dynamodb_table = "my-lock-table"
+    }   
+  }  
+}
+```
+
+You would migrate this to `terragrunt.hcl` as follows:
+
+```hcl
+# terragrunt.hcl
+
+terraform {
+  source = "git::git@github.com:foo/modules.git//frontend-app?ref=v0.0.3"
+
+  extra_arguments "custom_vars" {
+    commands  = ["apply", "plan"]
+    arguments = ["-var", "foo=42"]
+  }
+}
+
+remote_state {
+  backend = "s3"
+  config = {
+    bucket         = "my-terraform-state"
+    key            = "${path_relative_to_include()}/terraform.tfstate"
+    region         = "us-east-1"
+    encrypt        = true
+    dynamodb_table = "my-lock-table"
+  }   
+}  
+```
+
+
+### Move input variables into inputs
+
+When we were using `terraform.tfvars` files for Terragrunt configuration, we were piggybacking on the fact that 
+Terraform [automatically loads variables from tfvars 
+files](https://www.terraform.io/docs/configuration/variables.html#variable-definitions-tfvars-files) to set variables
+for our modules:
+
+```hcl
+# terraform.tfvars
+
+# Terragrunt configuration
+terragrunt = {
+  terraform {
+    # ...
+  }
+  
+  remote_state {
+    # ...
+  }
+}
+
+# Input variables to set for your Terraform module
+instance_type  = "t2.micro"
+instance_count = 10
+``` 
+
+With the move to `terragrunt.hcl`, we no longer get this behavior for free. However, Terragrunt can simulate this 
+behavior for you if you define your input variables by specifying `inputs = { ... }`:
+
+```hcl
+# terragrunt.hcl
+
+terraform {
+  # ...
+}
+
+remote_state {
+  # ...  
+}
+
+# Input variables to set for your Terraform module
+inputs = {
```

Comment:
The block vs attr syntax is a bit confusing, but either hcl2 or terraform 0.12 impose some validation on this distinction. Just wanting to clarify here that `terraform` and `remote_state` truly are blocks, where `inputs` truly is an attr. Meaning this would not work (with an =):

```
terraform = {
  # ...
}

remote_state = {
  # ...  
}
```

and notionally, because those two are blocks, you could have several of them in the same config, somehow?

Reply:
I found the HCL2 parser was a bit weird with these. To be able to define the attributes within these `terraform` and `remote_state`, I had to make them blocks... They really should only appear once, so I wanted to make them attributes, but I got confusing errors when I did that. Maybe attributes can't have well-defined properties within? Not sure.

Reply:
Yeah, I'm frustrated with this change in hcl2 at the moment. Not sure if you've seen it already, but there is a property you can set on the schema to treat a block as an attr. The many, many PRs linked to [this issue](https://github.com/hashicorp/terraform/issues/20505#issuecomment-496603984) might point to a way forward to use attrs?

Reply:
Here is an example: https://github.com/terraform-providers/terraform-provider-aws/pull/8228/files

Reply:
Also just found the behavior better described here, https://www.terraform.io/docs/extend/terraform-0.12-compatibility.html#computed-resource-attributes